### ISSUE TITLE
Support for Bayesian Convolutional Networks

### ISSUE THREAD

**Issue by crockpotveggies** [eclipse/deeplearning4j#1527]
Feature request for adding Bayesian ConvNets.

As per the abstract from [Bayesian Convolutional Networks with Bernoulli Approximate Variational Inference](http://mlg.eng.cam.ac.uk/yarin/PDFs/NIPS_2015_bayesian_convnets.pdf):

> Convolutional neural networks (convnets) work well on large datasets. But labelled data is hard to collect, and in some applications larger amounts of data are not available. The problem then is how to use convnets with small data – as convnets overfit quickly. We present an efficient Bayesian convnet, offering better robustness to over-fitting on small data than traditional approaches. This is by placing a probability distribution over the convnet’s kernels. To make this possible we present new theoretical results casting dropout network training as approximate inference in Bayesian neural networks. This allows us to implement our model using existing tools in the field with no increase in time complexity. We approximate our model’s intractable posterior with Bernoulli variational distributions, requiring no additional model parameters. We show a considerable improvement in classification accuracy compared to standard techniques with state-of-the-art results on CIFAR-10.

**Comment by velraman** [comment]
Hi,
Can I start working in this ?

Thanks

**Comment by AlexDBlack** [comment]
@velraman AFAIK nobody is currently working on this - so feel free.
Also fyi there's no shortage of other architectures that we could implement in DL4J too - some of them might not have issues. Feel free to ask if you want details.

**Comment by zhangy10** [comment]
Hey @AlexDBlack, I am also wondering if the Bayesian Convolutional Networks has been added in the latest beta version? I tried to search for it in the codes but I didn't find it.

**Comment by AlexDBlack** [comment]
@zhangy10 No, unfortunately we haven't added any functionality related to Bayesian neural networks, and it's not on our immediate roadmap.

**Comment by agibsonccc** [comment]
Closing stale issue: cleanup.

@zhangy10  I would suggest if we're missing anything like this, we're focusing a lot more on model import. It might be found via model import if we do it. Look at our samediff framework for implementing new features like this in the future.
